### PR TITLE
Bring over number utils from wallet and explorer

### PR DESCRIPTION
**Pull request by nop33**
Related to #35

### CODE REVIEW

**Review by mvaivre — APPROVED** [review]

File: test/numbers-test.ts [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+  // expect(countDecimals(10000000000.0000001)).toEqual(7),
+  // expect(countDecimals(1000000000.00000001)).toEqual(8),
+  // expect(countDecimals(100000000.000000001)).toEqual(9),
+  // expect(countDecimals(10000000.0000000001)).toEqual(10),
+  // expect(countDecimals(1000000.00000000001)).toEqual(11),
+  // expect(countDecimals(100000.000000000001)).toEqual(12),
+  // expect(countDecimals(10000.0000000000001)).toEqual(13),
+  // expect(countDecimals(1000.00000000000001)).toEqual(14),
+  // expect(countDecimals(100.000000000000001)).toEqual(15),
+  // expect(countDecimals(10.0000000000000001)).toEqual(16)
```
Comment (polarker) [inline test/numbers-test.ts:210]: Are these dead code expected?

File: lib/numbers.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  if (amount.includes('e')) {
+    cleanedAmount = convertScientificToFloatString(amount)
+  }
+
+  const numberOfDecimals = cleanedAmount.includes('.') ? cleanedAmount.length - 1 - cleanedAmount.indexOf('.') : 0
+  const numberOfZerosToAdd = 18 - numberOfDecimals
+  return BigInt(`${cleanedAmount.replace('.', '')}${produceTrailingZeros(numberOfZerosToAdd)}`)
+}
+
+export const convertScientificToFloatString = (scientificNumber: string): string => {
```
Comment (polarker) [inline lib/numbers.ts:144]: There is no built-in support for this, any mature library to do this? Scientific number has it's own standard. Writing this manually usually does not comply the standard

**Comment by polarker** [comment]
Could you guys check how it is done by Metamask or other popular projects?

File: test/numbers-test.ts (hunk 1 above)
Comment (nop33) [reply test/numbers-test.ts:210]: I have discovered these edge cases where JS has precision issues. I haven't worked on solving them yet, so I added them as a comment for when I will find time to extend the tests.

**Comment by nop33** [comment]
Metamask seems to be using 2 libraries (based on a [file about conversions](https://github.com/MetaMask/metamask-extension/blob/develop/shared/modules/conversion.utils.js)): [bignumber.js](https://www.npmjs.com/package/bignumber.js) and [BN.js](https://www.npmjs.com/package/bn.js) (through re-exporting it from [ethereum-utils](https://www.npmjs.com/package/ethereumjs-util))

**Comment by polarker** [comment]
> Metamask seems to be using 2 libraries (based on a [file about conversions](https://github.com/MetaMask/metamask-extension/blob/develop/shared/modules/conversion.utils.js)): [bignumber.js](https://www.npmjs.com/package/bignumber.js) and [BN.js](https://www.npmjs.com/package/bn.js) (through re-exporting it from [ethereum-utils](https://www.npmjs.com/package/ethereumjs-util))

I guess these libs are for the underlying representation, not for the UI display. Not sure if metamask use any lib for the display of numbers

**Comment by nop33** [comment]
I've spent a considerable amount of time trying to figure this out, with no definitive result.

image

The code that corresponds to the element in the screenshot below [is here](https://github.com/MetaMask/metamask-extension/blob/09d81ac5f2c94d7730d2e981b4f05734858030bf/ui/components/ui/currency-display/currency-display.component.js#L41). Through some [digging](https://github.com/MetaMask/metamask-extension/blob/09d81ac5f2c94d7730d2e981b4f05734858030bf/ui/hooks/useCurrencyDisplay.js#L53) I found that the [`displayValue` is set here](https://github.com/MetaMask/metamask-extension/blob/b5e601bc872d88358dab159d735ea9b32cb140b8/ui/components/app/wallet-overview/token-overview.js#L68), which led me to the [`balanceToRender`](https://github.com/MetaMask/metamask-extension/blob/b5e601bc872d88358dab159d735ea9b32cb140b8/ui/components/app/wallet-overview/token-overview.js#L48), which led me to the place where [`tokensWithBalances ` is is set](https://github.com/MetaMask/metamask-extension/blob/b5e601bc872d88358dab159d735ea9b32cb140b8/ui/hooks/useTokenTracker.js#L39), which led me to the place where the [`updateBalances` event is emitted](https://github.com/MetaMask/eth-token-tracker/blob/87fd6d6057559cac7cf066faddeeb4a5e45e641a/lib/index.js#L107), which sent me looking for the place where `token.update()` is defined to find out what it returns, which led me [here](https://github.com/MetaMask/metamask-extension/blob/1b6e58c41766c488ebb26d252e409278a5de177b/ui/pages/asset/asset.js#L14), which led me [here](https://github.com/MetaMask/metamask-extension/blob/1b6e58c41766c488ebb26d252e409278a5de177b/ui/ducks/metamask/metamask.js#L257)...

I can't find any library that is used for displaying balances in the UI...

**Comment by mvaivre** [comment]
So it seems they don't use any library. That's why they even display raw scientific notation for big numbers, as seen in the issue mentioned earlier...!

Our approach represents a good mix between improved readability (hence usability), and precision. Let's not forget that full precision is available in the explorer, which is one click away. And it will be available directly in the wallet in an upcoming release, in a "detail panel" as seen [in my mockups](https://www.figma.com/proto/wDWcIxJIq5aBTLwMPGb1Js/Alephium-Desktop-Wallet---DeFi-evolution?page-id=0%3A1&node-id=1%3A2&viewport=241%2C48%2C0.79&scaling=min-zoom&starting-point-node-id=1%3A2) (click on a transaction).

I believe our approach is sensible. We'll obviously follow our user feedback, and adapt it if necessary. 

@polarker Can we proceed and merge this? Thanks for your feedback!

File: lib/numbers.ts [hunk 3]
```diff
@@ -47,44 +47,42 @@ export const getNumberOfTrailingZeros = (numberArray: string[]) => {
...
 
   // For abbreviation, we don't need full precision and can work with number
   const alephNum = Number(baseNum) / QUINTILLION
 
-  // what tier? (determines SI symbol)
-
-  let tier = (Math.log10(alephNum) / 3) | 0
-
-  const numberOfDigitsToDisplay = nbOfDecimalsToShow || minDigits
+  const minNumberOfDecimals = alephNum > -0.01 && alephNum < 0.01 ? 3 : 2
```
Comment (polarker) [inline lib/numbers.ts:69]: the case `alephNum < 0` is handled on line 64

File: lib/numbers.ts [hunk 4]
```diff
@@ -47,44 +47,42 @@ export const getNumberOfTrailingZeros = (numberArray: string[]) => {
...
+  const tinyAmountsMaxNumberDecimals = 5
+  const numberOfDecimalsToDisplay = nbOfDecimalsToShow || minNumberOfDecimals
 
-  if (tier < 0 || showFullPrecision) {
-    // Keep full precision for very low numbers (gas etc.)
+  if (showFullPrecision) {
```
Comment (polarker) [inline lib/numbers.ts:73]: Let's move this if statement before line 70 as it does not count on line70, 71

File: lib/numbers.ts [hunk 5]
```diff
@@ -47,44 +47,42 @@ export const getNumberOfTrailingZeros = (numberArray: string[]) => {
...
+    const decimals = countDecimals(alephNum) === 1 ? 16 : 18 // Avoid precision issue edge case
+    return removeTrailingZeros(alephNum.toFixed(decimals), minNumberOfDecimals)
+  }
 
-    if (countDecimals(alephNum) === 1) {
-      return removeTrailingZeros(alephNum.toFixed(16), minDigits) // Avoid precision issue edge case
-    }
+  let tier = (Math.log10(alephNum) / 3) | 0 // what tier? (determines SI symbol)
```
Comment (polarker) [inline lib/numbers.ts:78]: Using `Math.log10` is not efficient, we could compare directly with `0.001, 1000000, etc`

File: test/numbers-test.ts [hunk 6]
```diff
@@ -35,20 +35,20 @@ const minDigits = 3
...
-    expect(abbreviateAmount(BigInt('2000000000000000000'))).toEqual('2.000'),
-    expect(abbreviateAmount(alf(BigInt(1230)))).toEqual('1.230K'),
-    expect(abbreviateAmount(alf(BigInt(1230000)))).toEqual('1.230M'),
-    expect(abbreviateAmount(alf(BigInt(1230000000)))).toEqual('1.230B'),
-    expect(abbreviateAmount(alf(BigInt(1230000000000)))).toEqual('1.230T'),
-    expect(abbreviateAmount(alf(BigInt(1230000000000000)))).toEqual('1230.000T'),
-    expect(abbreviateAmount(alf(BigInt(1)))).toEqual('1.000')
+    expect(abbreviateAmount(BigInt('2000000000000000000'))).toEqual('2.00'),
+    expect(abbreviateAmount(alf(BigInt(1230)))).toEqual("1'230.00"),
+    expect(abbreviateAmount(alf(BigInt(1230000)))).toEqual('1.23M'),
```
Comment (polarker) [inline test/numbers-test.ts:47]: need a test case for `alf(BigInt(1234000) == 1.23M`

**Review by polarker — COMMENTED** [review]
A few more comments. Let's not force push changes for critical code

File: lib/numbers.ts (hunk 3 above)
Comment (nop33) [reply lib/numbers.ts]: fixed

File: lib/numbers.ts (hunk 4 above)
Comment (nop33) [reply lib/numbers.ts:70]: fixed

File: lib/numbers.ts (hunk 5 above)
Comment (nop33) [reply lib/numbers.ts]: fixed

File: test/numbers-test.ts (hunk 6 above)
Comment (nop33) [reply test/numbers-test.ts:47]: fixed

File: lib/numbers.ts [hunk 7]
```diff
@@ -65,23 +65,33 @@ export const abbreviateAmount = (baseNum: bigint, showFullPrecision = false, nbO
...
+      : 5
 
   if (tier < 0) {
-    // amount <= 0.001
     return removeTrailingZeros(alephNum.toFixed(tinyAmountsMaxNumberDecimals), minNumberOfDecimals)
   } else if (tier <= 1) {
-    // amount < 1'000'000
     return addApostrophe(removeTrailingZeros(alephNum.toFixed(numberOfDecimalsToDisplay), minNumberOfDecimals))
   } else if (tier >= MONEY_SYMBOL.length) {
     tier = MONEY_SYMBOL.length - 1
```
Comment (polarker) [inline lib/numbers.ts:97]: We could remove the computation of `tier` as follows:
```
if (alephNum < 0.001) {
  return ...
} else if (alephNum < 1000000) {
  return ...
}

// last case
let tier = MONEY_SYMBOL.length - 1
```

**Comment by polarker** [comment]
Last comment: we should replace all `aleph` with `alph` now

File: lib/numbers.ts (hunk 7 above)
Comment (nop33) [reply lib/numbers.ts:97]: That would break the code.

We only want to override `tier` when its value is `>= 5`. Otherwise, if I do what you suggest, we will override `tier` even when it's `2`, `3`, or `4`, which is not what we want.

Comment (polarker) [reply lib/numbers.ts:95]: that was for demostration, not correct code tho. To be more precise:
```
// last case
let tier = alephNum < 1000000000
      ? 2
      : alephNum < 1000000000000
      ? 3
      : alephNum < 1000000000000000
      ? 4
      : 5
if (tier >= MONEY_SYMBOL.length) {
  tier = MONEY_SYMBOL.length - 1
}
```
In this way, for most of the cases, you don't need to compute `tier`

Comment (nop33) [reply lib/numbers.ts:95]: Actually, the total ALPH supply is 1'000'000'000 (tier 4). We don't even need to worry about anything higher than that cause it's physically impossible.

Comment (polarker) [reply lib/numbers.ts:95]: perfect 👍

**Comment by nop33** [comment]
@polarker fixed all your comments

**Comment by polarker** [comment]
@nop33 How does this commit sound to you? https://github.com/alephium/alephium-js/commit/bee3f9f30c5b3210e25c476f7f49548066d7a9b6

**Comment by nop33** [comment]
@polarker done, thanks!

**Review by polarker — APPROVED** [review]
@nop33 Thanks for enduring my nitpicking. We are ready to merge it now
